### ISSUE TITLE
BSD service: Recv buffer size 0

### ISSUE THREAD

**Issue by StephenCWills** [Ryujinx/Ryujinx#2393]
### What's the issue you encountered?

In calls to `recv()` from a homebrew app, the BSD service in Ryujinx is attempting to write to a zero-length buffer.

### Logging in the homebrew app:

Here is the relevant code from the homebrew app (C++).
```C++
fprintf(stderr, "recv1(%d)\n", static_cast<int>(size));
signed_size_type result = ::recv(s, static_cast<char*>(data), size, flags);
```

The `fprintf()` call above produced the following message in Ryujinx.
```
00:00:46.897 |W| HLE.GuestThread.15 KernelSvc OutputDebugString: recv1(65535)
```

### Screenshots from debugging Ryujinx:

This is a screenshot from [IClient.cs](https://github.com/Ryujinx/Ryujinx/blob/50ba233ac628ffb9661ea2bda2cda216d18e81c1/Ryujinx.HLE/HOS/Services/Sockets/Bsd/IClient.cs#L470) that shows the size of the buffer after it's been retrieved via `GetBufferType0x22()`.
> ![image](https://user-images.githubusercontent.com/9203145/123018340-5c30e800-d39c-11eb-9919-8969ef499b7b.png)

Here is a screenshot from when I stepped into `GetBufferType0x22()` showing which branch of the logic it took.
> ![image](https://user-images.githubusercontent.com/9203145/123017844-6bfbfc80-d39b-11eb-8207-f33c9391b479.png)

I also checked the size of `ReceiveBuff`, and it seems to match the size of the buffer passed into `::recv()`.
> ![image](https://user-images.githubusercontent.com/9203145/123017848-6ef6ed00-d39b-11eb-85ad-e2d520291d1c.png)

### How can the issue be reproduced?

The steps I've used to reproduce this are fairly involved since it requires setup of a multiplayer game in [DevilutionX](https://github.com/diasurgical/devilutionX). To reproduce, you'll need to use a test build of the game linked to this PR: https://github.com/diasurgical/devilutionX/pull/2199. I had to host a game on my PC using the Windows x64 build and then attempt to join it using the Switch build via Ryujinx.

### Additional context?

I tried digging into this [IPC stuff](https://switchbrew.org/wiki/IPC_Marshalling) myself, but it's making my head spin. As far as I can tell, Ryujinx is trying to write data from the socket into the "ReceiveList" described by buffer descriptor C when maybe it's supposed to be writing it into the "Receive" buffer described by buffer descriptor B. Perhaps the distinction between buffer descriptors B and C is meant to mirror the distinction between the `recv()` and `recvmsg()` function calls in C++? I'm hoping someone else with a better understanding of this can help figure out what's supposed to be happening.

**Comment by StephenCWills** [comment]
FYI, this was just an experiment of mine, but it ended up working after I tried switching to using `ReceiveBuff` instead of `RecvListBuff`. I don't really understand what `GetBufferType0x22()` is supposed to represent so I'm not sure whether this is a good solution in general.

```C#
[CommandHipc(8)]
// Recv(u32 socket, u32 flags) -> (i32 ret, u32 bsd_errno, array<i8, 0x22> message)
public ResultCode Recv(ServiceCtx context)
{
    int         socketFd    = context.RequestData.ReadInt32();
    SocketFlags socketFlags = (SocketFlags)context.RequestData.ReadInt32();

    ulong receivePosition = 0;
    ulong receiveLength = 0;

    if (context.Request.ReceiveBuff.Count > 0)
    {
        receivePosition = context.Request.ReceiveBuff[0].Position;
        receiveLength = context.Request.ReceiveBuff[0].Size;
    }

    ...
```

![image](https://user-images.githubusercontent.com/9203145/123097326-0d1b9f00-d3fe-11eb-9904-ea34a61d6a8c.png)

**Comment by StephenCWills** [comment]
It took me a bit, but I got one of our Switch players to test the game on the console itself. He reported the same issue that I encountered when Ryujinx was attempting to write to a zero-length buffer. That leads me to believe there is an issue in libnx rather than Ryujinx.

I've been digging into the libnx code and reading and rereading the IPC Marshalling page, and I believe I now understand what `GetBufferType0x22()` is doing. I'm pretty sure the issue I'm having is related to the following statement from the IPC Marshalling page.

> For buffers with flag 0x20 it creates two descriptors (A+X or B+C), **but one descriptor is NULL (zero size and pointer)**, while the other holds the expected values.

Basically, the reason that Ryujinx and the Switch console are selecting the wrong buffer is because buffer descriptor C has a non-zero pointer, when it is presumably supposed to be zero. I'm going to dig into libnx a bit longer and see about bringing this up with the libnx developers to see what they say.

**Comment by StephenCWills** [comment]
Working with one of the devkitPro developers, we tested a simple example that confirms the issue here is unrelated to the issue that our tester is dealing with in DevilutionX. Rather, I believe I found the cause for this issue the other day, and it starts in `KServerSession.Receive()`.

https://github.com/Ryujinx/Ryujinx/blob/f4078ae2670bf98ce0e9e41359a3614ac802946f/Ryujinx.HLE/HOS/Kernel/Ipc/KServerSession.cs#L320-L321

The first two words of the client's IPC message are copied to server process memory, which includes the flags that define the number of C buffer descriptors in the message. However, there is no code in `KServerSession.Receive()` that actually handles the client's C buffer descriptors. Instead, whatever data is already sitting in server memory at the time will remain in the space where the C buffer descriptors should have been defined. This discrepancy is later remedied with a bit of code in `ServerBase` that overwrites the parsed C buffer descriptors when the IPC message is processed.

https://github.com/Ryujinx/Ryujinx/blob/f4078ae2670bf98ce0e9e41359a3614ac802946f/Ryujinx.HLE/HOS/Services/ServerBase.cs#L177-L196

The problem here is that the `noReceive` flag is set to `false` if `ReceiveBuff` has a nonzero `Position`, as is the case with 0x22 buffers that are too big to fit in the pointer buffer. That causes `ServerBase` to skip fixing the `ReceiveListBuff` pointers and so the IPC message is ultimately handled using a `ReceiveListBuff` collection that was parsed from whatever happened to be in server memory where the C buffer descriptors should have been defined.
